sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/handler/XdsHandler.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2024-2024 Sermant Authors. All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package io.sermant.implement.service.xds.handler;
+
+import io.envoyproxy.envoy.config.core.v3.Node;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryRequest;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryResponse;
+import io.grpc.stub.StreamObserver;
+import io.sermant.core.common.LoggerFactory;
+import io.sermant.core.config.ConfigManager;
+import io.sermant.core.plugin.config.ServiceMeta;
+import io.sermant.core.utils.NetworkUtils;
+import io.sermant.implement.service.xds.client.XdsClient;
+import io.sermant.implement.service.xds.env.XdsConstant;
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * abstract xds protocol handler
+ *
+ * @author daizhenyu
+ * @since 2024-05-13
+ **/
+public abstract class XdsHandler implements XdsServiceAction {
+    protected static final Logger LOGGER = LoggerFactory.getLogger();
+
+    protected final XdsClient client;
+
+    protected final String resourceType;
+
+    protected Node node;
+
+    /**
+     * construction method
+     *
+     * @param client xds client
+     * @param resourceType xds resource type
+     */
+    public XdsHandler(XdsClient client, String resourceType) {
+        this.client = client;
+        this.resourceType = resourceType;
+        createNode();
+    }
+
+    /**
+     * built DiscoveryRequest
+     *
+     * @param type resource type
+     * @param version resource version
+     * @param nonce response nonce
+     * @param resourceName resource name
+     * @return DiscoveryRequest
+     */
+    protected DiscoveryRequest builtDiscoveryRequest(String type, String version, String nonce,
+            List<String> resourceName) {
+        DiscoveryRequest.Builder builder = DiscoveryRequest.newBuilder()
+                .setNode(node)
+                .setTypeUrl(type);
+        if (version != null) {
+            builder.setVersionInfo(version);
+        }
+        if (nonce != null) {
+            builder.setResponseNonce(nonce);
+        }
```

Comment:
it is better to set default version and response nonce when null

Reply:
no default version and response nonce. the version and nonce getted from response to send ack.

Reply:
what if there exists no version  and response nonce

Reply:
When you actively subscribe to a resource